File: src/com/medium/BinaryTreeInorderTraversal_94.java
/**
 * 
 */
package com.medium;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class BinaryTreeInorderTraversal_94 {

    private List<Integer> mAnswer = new ArrayList<>();

    // Recursive version (LDR)
    // Time Complexity: O(n)
    // Space Complexity: O(h)
    // Runtime: 0 ms, faster than 100.00%
    // Memory Usage: 35 MB, less than 99.99%
    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null) {
            return mAnswer;
        }
        inorderTraversal(root.left);
        mAnswer.add(root.val);
        inorderTraversal(root.right);

        return mAnswer;
    }

    // Iterative version (LDR)
    // Time Complexity: O(n)
    // Space Complexity: O(h)
    // Runtime: 1 ms, faster than 53.57%
    // Memory Usage: 34.7 MB, less than 99.99%
    public List<Integer> inorderTraversal2(TreeNode root) {
        if (root == null) {
            return mAnswer;
        }
        Stack<TreeNode> mStack = new Stack<>();
        TreeNode mCurrent = root;
        while (mCurrent != null || !mStack.isEmpty()) {
            while (mCurrent != null) {
                mStack.push(mCurrent);
                mCurrent = mCurrent.left;
            }

            TreeNode mNode = mStack.pop();

            mAnswer.add(mNode.val);
            mCurrent = mNode.right;
        }

        return mAnswer;
    }
}

File: Golang_src/medium/binaryTreeInorderTraversal_94.go
package medium

// TreeNode for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Use recursive:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2.1 MB, less than 22.04%
func inorderTraversal(root *TreeNode) []int {
	if root == nil {
		return nil
	}
	inorder := []int{}
	inorder = append(inorder, inorderTraversal(root.Left)...)
	inorder = append(inorder, root.Val)
	inorder = append(inorder, inorderTraversal(root.Right)...)

	return inorder
}

// Use iterative:
// Time Complexity: O(n)
// Space Complexity:O(n)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 22.04%
func inorderTraversal2(root *TreeNode) []int {
	stack := []*TreeNode{}
	result := []int{}

	cur := root

	for cur != nil || len(stack) > 0 {
		for cur != nil {
			stack = append(stack, cur)
			cur = cur.Left
		}
		cur = stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		result = append(result, cur.Val)
		cur = cur.Right
	}

	return result
}
